src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -1,60 +1,1057 @@
+        ctx.Definitions
+        |> List.map (fun def -> def.SelectionSet)
+        |> List.fold (fun acc selectionSet -> checkArgumentUniqueness acc selectionSet) Success
+
+    let private checkRequiredArguments (acc : ValidationResult) (schemaInfo : SchemaInfo) (selection : SelectionInfo) =
+        let acc =
+            selection.InputValues
+            |> Option.map (Array.fold (fun acc argDef ->
+                match argDef.Type.Kind with
+                | TypeKind.NON_NULL when argDef.DefaultValue.IsNone ->
+                    match selection.Field.Arguments |> List.tryFind (fun arg -> arg.Name = argDef.Name) with
+                    | Some arg when arg.Value <> EnumValue "null" -> acc // TODO: null values are being mapped into an enum value! Isn't it better to have a case for null values?
+                    | _ -> acc @ Error.AsResult(sprintf "Argument '%s' of field '%s' of type '%s' is required and does not have a default value." argDef.Name selection.Field.Name selection.FragmentOrParentType.Name, selection.Path)
+                | _ -> acc) acc)
+            |> Option.defaultValue acc
+        selection.Field.Directives
+        |> List.fold (fun acc directive ->
+            match schemaInfo.Directives |> Array.tryFind (fun d -> d.Name = directive.Name) with
+            | Some directiveType ->
+                directiveType.Args
+                |> Array.fold (fun acc argDef ->
+                    match argDef.Type.Kind with
+                    | TypeKind.NON_NULL when argDef.DefaultValue.IsNone ->
+                        match selection.Field.Arguments |> List.tryFind (fun arg -> arg.Name = argDef.Name) with
+                        | Some arg when arg.Value <> EnumValue "null" -> acc // TODO: null values are being mapped into an enum value! Isn't it better to have a case for null values?
+                        | _ -> acc @ Error.AsResult(sprintf "Argument '%s' of directive '%s' of field '%s' of type '%s' is required and does not have a default value." argDef.Name directiveType.Name selection.Field.Name selection.FragmentOrParentType.Name, selection.Path)
+                    | _ -> acc) acc
+            | None -> acc) acc
+
+    let internal validateRequiredArguments (ctx : ValidationContext) =
+        ctx.Definitions
+        |> List.fold (fun acc def -> def.SelectionSet |> List.fold (fun acc selection -> checkRequiredArguments acc ctx.Schema selection) acc) Success
+
+    let internal validateFragmentNameUniqueness (ctx : ValidationContext) =
+        ctx.FragmentDefinitions
+        |> List.filter (fun f -> f.Definition.Name.IsSome)
+        |> List.groupBy (fun f -> f.Definition.Name.Value)
+        |> List.choose (fun (name, frags) -> if frags.Length > 1 then Some (name, frags.Length) else None)
+        |> List.fold (fun acc (name, length) -> acc @ Error.AsResult(sprintf "There are %i fragments with name '%s' in the document. Fragment definitions must have unique names." length name)) Success
```

Comment:
Instead of repeatedly retraversing the list, we could fuse these together into a single fold with a map containing name occurance counts.

Reply:
Done.